kmail: fix calling ext. applications

eliasp opened the pull request:
KMail relies on quite a few external applications, such as:
- kaddressbook
- akonadiimportwizard
- accountwizard

Without those, quite a few core functionalities of KMail aren't available, such as:
- adding a new account to KMail
- using the addressbook for sending a mail
- importing existing PIM data

By providing `QStandardPaths::findExecutable()` the store path of said
components, this problem can be circumvented.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Make KMail work again

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

eliasp commented:
Split it up now into 2 separate commits, so the actual change becomes more clear and is not obscured by the move of `kmail.nix` → `kmail/default.nix` in 57f182c.

eliasp commented:
Rebased + conflicts solved.

SuperSandro2000 commented:
ping @ttuegel

ttuegel reviewed (commented):
Usually, we would add the paths in the wrapper. Is there a reason we must patch the source here, instead?

SuperSandro2000 commented:
ping @eliasp

eliasp commented:
> Usually, we would add the paths in the wrapper. Is there a reason we must patch the source here, instead?

No reason - I just wasn't aware of this way. Fixed that & threw away my little more complicated approach :smile:

SuperSandro2000 commented on pkgs/applications/kde/kmail.nix:
```diff
@@ -18,13 +18,21 @@ mkDerivation {
   propagatedUserEnvPkgs = [ kdepim-runtime kwallet akonadi ];
+  postFixup = ''
+    wrapProgram "$out/bin/kmail" \
+      --prefix PATH : "${lib.makeBinPath [ akonadi-import-wizard ]}" \
+      --prefix PATH : "${lib.makeBinPath [ kaddressbook ]}" \
+      --prefix PATH : "${lib.makeBinPath [ kleopatra ]}" \
+      --prefix PATH : "${lib.makeBinPath [ kmail-account-wizard ]}" \
+      --prefix PATH : "${lib.makeBinPath [ pim-data-exporter ]}"
```
```suggestion
      --prefix PATH : "${lib.makeBinPath [ akonadi-import-wizard kaddressbook kleopatra kmail-account-wizard pim-data-exporter ]}"
```

eliasp replied:
TIL & fixed!

eliasp commented:
> ```
> libsForQt5.kmail: Ask if output path share/doc (5.6M) could be split with outputs = [ ... "doc" ];
> ```

See 2e2ab21 - is that what you asked for?

SuperSandro2000 commented:
Not sure if bin and lib make sense here. You most likely need them anyway. I think $out and $doc are enough.

eliasp committed:
kmail: nixpkgs-fmt + sort items

eliasp committed:
kmail: fix calling ext. applications

KMail relies on quite a few external applications, such as:
- kaddressbook
- akonadiimportwizard
- accountwizard
- akonadiselftest

Without those, quite a few core functionalities of KMail aren't available, such as:
- adding a new account to KMail
- using the addressbook for sending a mail
- importing existing PIM data
- analyzing Akonadi issues

Fixed by adding those components to kmail's wrapper's PATH.

eliasp committed:
kmail: use multiple outputs

eliasp commented:
> Not sure if bin and lib make sense here. You most likely need them anyway. I think $out and $doc are enough.

Makes sense - it's fixed now. Thanks!

SuperSandro2000 commented on pkgs/applications/kde/kmail.nix:
```diff
@@ -19,13 +58,53 @@ mkDerivation {
+    kleopatra
+    pim-data-exporter
   ];
+  outputs = [ "doc" "out" ];
```
```suggestion
  outputs = [ "out" "doc" ];
```

I am pretty sure doc should be the default when the attr is used.

eliasp replied:
The [documentation](https://nixos.org/manual/nixpkgs/stable/#sec-multiple-outputs-) doesn't really tell much about how to use `outputs`…
If anyone has a deeper understanding of this than me, this might be worth a PR… I just tried to make sense of [`pkgs/build-support/setup-hooks/multiple-outputs.sh`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/setup-hooks/multiple-outputs.sh), but I feel like I miss quite some knowledge regarding the internals…

Updated `kmail.nix` according to the suggestion.

SuperSandro2000 replied:
The first one listed in outputs gets used by default. Most of the time it should be "out".

schmittlauch commented:
Is this still waiting on something? I can offer to test this on my system, although I cannot promise to hit all use cases that then call external programs.

eliasp commented:
IMHO nothing left to be done for merging here… I addressed all open issues so far.

schmittlauch commented:
While testing #127332 I discovered that this PR only resolves this issue for KMail when being run via `kmail`, but not when embedded as a port of `kontact`.
I could just apply these changes to the kontact derivation as well, but that introduces code redundancy and the need to maintain changes for 2 files.

Any ideas for a cleaner solution? One way would be to expose the external KMail dependencies through a `passthru? attribute and then use these in the kontact derivation. But corrently, kontact does not even use kmail as an input :confused:

eliasp commented:
Does Kontact embed KMail as KPart? Then my [first approach](https://github.com/NixOS/nixpkgs/commit/2284709f7769330b1703e43210addc2ec056199b) would probably have made this work.

schmittlauch commented:
Yes, that's how I understand it. The kontact derivation doesn't take kmail as an input, but takes `kparts` and then probably makes a run-time discovery (yay, impure) of what PIM components are available.

schmittlauch commented:
Yes, it is definitely using KParts to embed KMail: When I just built `kontact` from the commit specified above (based on unstable), it failed to load the KMail part with this error:

`Cannot load part for Mail.
Cannot load library /home/spiollinux/.nix-profile/lib/qt-5.15.2/plugins/kmailpart.so: (/nix/store/z3lf066csyma75624syd2vcklry8vqhd-glib-2.68.2/lib/libgobject-2.0.so.0: undefined symbol: g_memdup2)`

So for proper testing, I might need to overlay the whole KDE ecosystem. Or backport your changes.

schmittlauch commented:
Even if the patches proposed by @eliasp work, we need to weigh the effort of maintaining them with the drawbacks of the simple wrapper solution and wrapping both kmail and kontact separately.

Do you think your patches also work on other distros and are thus upstreamable? Because then this might be a clear reason to adopt that approach.

eliasp commented:
No, those patches are quite NixOS-specific (hardcoding the binary path for external applications called via `QStandardPaths::findExecutable()` to their store path), as on all other distros, said external applications should be located dynamically based on the runtime environment's `$PATH`.

Auto-patching all usages of `QStandardPaths::findExecutable()` during buildtime (even for all other Qt applications) won't work as well, as there's to proper way to always determine the application to be executed.

So wrapping `kmail` + `kontact` seperately seems to be most reasonable approach IMHO.
